File: vm.go
package main

import (
	"bytes"
	"fmt"
	"slices"
)

// type Opcode byte
// type BinaryOpcode byte
// TODO: don't want to be casting `byte(opcode)` all the time

const (
	OpPop byte = iota
	OpBinary
	OpNot
	OpJumpIfFalse
	OpJumpForward
	OpJumpBack
	OpInlineNumber
	OpLoadConstant
	OpReadVariable
	OpSetVariable
	OpInstantiate
	OpCallBuiltin
	OpCallFunction
	OpPrintln // Special op because it's variadic
	OpFieldAccess
	OpSetField
	OpDuplicate

	InvalidOp
)

const (
	MaxBlockSize = 60_000
	MaxConstants = 250
)

const (
	OpBinaryPlus byte = iota
	OpBinarySubtract
	OpBinaryMultiply
	OpBinaryDivide
	OpBinaryRemainder

	OpBinaryEqual
	OpBinaryGreaterThan
	OpBinaryLessThan

	OpBinaryBinaryOr
	OpBinaryBinaryXor
	OpBinaryBinaryAnd

	OpBinaryConcat
)

type VmType struct {
	Name   string
	Fields []string
}

type VmInstance struct {
	vmType *VmType
	values []any
}

func (instance *VmInstance) print(out *bytes.Buffer) {
	out.WriteString(instance.vmType.Name)
	out.WriteRune('{')
	for i := range len(instance.values) {
		fieldName := instance.vmType.Fields[i]
		fieldValue := instance.values[i]
		if i != 0 {
			out.WriteRune(',')
		}
		out.WriteString(fieldName)
		out.WriteRune('=')
		writeValue(fieldValue, out)
	}
	out.WriteRune('}')
}

type VmFunction struct {
	params              []string
	ops                 []byte
	variableDefinitions []string
	hasOutVar           bool
}

type Vm struct {
	ops                 []byte
	constants           []any
	variableDefinitions []string
	variables           []any
	functions           map[string]VmFunction
	types               map[string]VmType
}

const maxStack = 50

func execute(ops []byte, constants []any, variableDefinitions []string, functions map[string]VmFunction, types map[string]VmType) error {
	variables := make([]any, len(variableDefinitions))
	vm := &Vm{
		ops:                 ops,
		constants:           constants,
		functions:           functions,
		types:               types,
		variableDefinitions: variableDefinitions,
		variables:           variables,
	}
	stack := make([]any, maxStack)
	err := vm.execute(stack)
	return err
}

func (vm *Vm) execute(stack []any) error {
	constants, ops, functions, types := vm.constants, vm.ops, vm.functions, vm.types

	ip := 0
	readOpByte := func() byte {
		op := ops[ip]
		ip++
		return op
	}
	getConstant := func(index int) (string, error) {
		constant := constants[index]
		constantString, ok := constant.(string)
		if !ok {
			return "", fmt.Errorf("expected constant %d to be a string, but was '%v' at %d", index, constant, ip)
		}
		return constantString, nil
	}
	readConstantString := func() (string, error) {
		return getConstant(int(readOpByte()))
	}

	stackNext := 0
	popStack := func() any {
		stackNext -= 1
		return stack[stackNext]
	}
	pushStack := func(v any) {
		if stackNext == maxStack {
			panic(fmt.Sprintf("stack overflow: attempting to push '%v' onto the stack with maximum size %d", v, maxStack))
		}
		stack[stackNext] = v
		stackNext += 1
	}

	for ip < len(ops) {
		instruction := readOpByte()

		switch instruction {
		case OpPop:
			_ = popStack()
		case OpBinary:
			binop := readOpByte()
			right := popStack()
			left := popStack()

			var result any
			var err error

			switch binop {
			case OpBinaryPlus:
				result, err = intBinaryOp(left, right, "+", func(l int, r int) int { return l + r })
			case OpBinarySubtract:
				result, err = intBinaryOp(left, right, "-", func(l int, r int) int { return l - r })
			case OpBinaryMultiply:
				result, err = intBinaryOp(left, right, "*", func(l int, r int) int { return l * r })
			case OpBinaryDivide:
				result, err = intBinaryOp(left, right, "/", func(l int, r int) int { return l / r })
			case OpBinaryRemainder:
				result, err = intBinaryOp(left, right, "%", func(l int, r int) int { return l % r })
			case OpBinaryBinaryAnd:
				result, err = intBinaryOp(left, right, "%", func(l int, r int) int { return l & r })
			case OpBinaryBinaryOr:
				result, err = intBinaryOp(left, right, "%", func(l int, r int) int { return l | r })
			case OpBinaryBinaryXor:
				result, err = intBinaryOp(left, right, "%", func(l int, r int) int { return l ^ r })

			case OpBinaryEqual:
				result = boolToInt(isEqual(left, right))
			case OpBinaryGreaterThan:
				result, err = intBinaryOp(left, right, ">", func(l int, r int) int { return boolToInt(l > r) })
			case OpBinaryLessThan:
				result, err = intBinaryOp(left, right, "<", func(l int, r int) int { return boolToInt(l < r) })

			case OpBinaryConcat:
				result, err = stringConcat(left, right)

			default:
				return fmt.Errorf("unsupported binary operator %v at %d", binop, ip)
			}

			if err != nil {
				return err
			}

			pushStack(result)
		case OpNot:
			v := popStack()
			i, ok := v.(int)
			if !ok {
				return fmt.Errorf("operand of NOT operation must be int, but was '%v' at %d", v, ip)
			}
			pushStack(boolToInt(!intToBool(i)))
		case OpJumpIfFalse:
			b1 := int(readOpByte())
			b2 := int(readOpByte())
			jumpAmount := b1*256 + b2
			v := popStack()
			if !isWeirdlyTrue(v) {
				ip += jumpAmount
			}
		case OpJumpForward:
			b1 := int(readOpByte())
			b2 := int(readOpByte())
			jumpAmount := b1*256 + b2
			ip += jumpAmount
		case OpJumpBack:
			b1 := int(readOpByte())
			b2 := int(readOpByte())
			jumpAmount := b1*256 + b2
			ip -= jumpAmount
		case OpInlineNumber:
			v := int(readOpByte())
			pushStack(v)
		case OpLoadConstant:
			index := int(readOpByte())
			pushStack(constants[index])
		case OpReadVariable:
			index := int(readOpByte())
			value := vm.variables[index]
			if value == nil {
				variableName := vm.variableDefinitions[index]
				return fmt.Errorf("variable '%v' not defined at %d", variableName, ip)
			}
			pushStack(value)
		case OpSetVariable:
			index := int(readOpByte())
			vm.variables[index] = popStack()
		case OpInstantiate:
			typeName, err := readConstantString()
			if err != nil {
				return err
			}
			vmType, found := types[typeName]
			if !found {
				return fmt.Errorf("type '%v' not found at %d", typeName, ip)
			}
			fieldValues := make([]any, len(vmType.Fields))
			for i := range vmType.Fields {
				fieldValues[i] = popStack()
			}
			slices.Reverse(fieldValues) // Arguments were pushed onto the stack in left-to-right order, so we read them right-to-left
			instance := VmInstance{
				vmType: &vmType,
				values: fieldValues,
			}
			pushStack(&instance)
		case OpCallBuiltin:
			functionName, err := readConstantString()
			if err != nil {
				return err
			}
			builtin, found := builtins[functionName]
			if !found {
				return fmt.Errorf("builtin function '%v' not found at %d", functionName, ip)
			}
			arguments := make([]any, builtin.Arity)
			for i := 0; i < builtin.Arity; i++ {
				arguments[i] = popStack()
			}
			slices.Reverse(arguments) // Arguments were pushed onto the stack in left-to-right order, so we read them right-to-left
			returnValue, err := builtin.VmFunc(arguments)
			if err != nil {
				return err
			}
			pushStack(returnValue)
		case OpCallFunction:
			functionName, err := readConstantString()
			if err != nil {
				return err
			}
			function := functions[functionName]
			functionVariables := make([]any, len(function.variableDefinitions))
			for i := len(function.params) - 1; i >= 0; i-- {
				functionVariables[i] = popStack()
			}

			functionVm := &Vm{
				ops:                 function.ops,
				constants:           constants,
				functions:           functions,
				variables:           functionVariables,
				variableDefinitions: function.variableDefinitions,
			}

			err = functionVm.execute(stack[stackNext:])
			if err != nil {
				return err
			}

			var outVar any = nil
			if function.hasOutVar {
				// E.g. if a function has 2 input parameters, and 1 output parameter, then the variable spot for the
				// output parameter is right after the input parameters, i.e. in the 3rd spot, or index 2, which is the
				// length of the params slice
				outVar = functionVariables[len(function.params)]
			}

			pushStack(outVar)
		case OpPrintln:
			argumentCount := int(readOpByte())
			arguments := make([]any, argumentCount)
			for i := 0; i < argumentCount; i++ {
				arguments[i] = popStack()
			}
			slices.Reverse(arguments) // Arguments were pushed onto the stack in left-to-right order, so we read them right-to-left
			returnValue, err := builtinPrintlnVm(arguments)
			if err != nil {
				return err
			}
			pushStack(returnValue)
		case OpFieldAccess:
			identifier, err := readConstantString()
			if err != nil {
				return err
			}
			target := popStack()
			instance, ok := target.(*VmInstance)
			if !ok {
				return fmt.Errorf("left-hand operand of '.' must be a type instance but was '%v'", target)
			}
			fieldFound := false
			for i, field := range instance.vmType.Fields {
				if field == identifier {
					pushStack(instance.values[i])
					fieldFound = true
					break
				}
			}
			if !fieldFound {
				return fmt.Errorf("field '%v' not found on type '%v'", identifier, instance.vmType.Name)
			}
		case OpSetField:
			identifier, err := readConstantString()
			if err != nil {
				return err
			}
			value := popStack()
			target := popStack()
			instance, ok := target.(*VmInstance)
			if !ok {
				return fmt.Errorf("left-hand operand of '.' must be a type instance but was '%v'", target)
			}
			fieldFound := false
			for i, field := range instance.vmType.Fields {
				if field == identifier {
					instance.values[i] = value
					fieldFound = true
					break
				}
			}
			if !fieldFound {
				return fmt.Errorf("field '%v' not found on type '%v'", identifier, instance.vmType.Name)
			}
		case OpDuplicate:
			v := popStack()
			pushStack(v)
			pushStack(v)

		default:
			return fmt.Errorf("unknown instruction %v at %d", instruction, ip)
		}
	}

	return nil
}

File: toirs/src/vm.rs
use crate::{Constant, FunctionDefinition};

pub fn run(
    instructions: &[u8],
    constants: &[Constant],
    variable_names: &[String],
    functions: &[FunctionDefinition],
) {
    let variables: Vec<i64> = vec![0; variable_names.len()];
    let mut stack = Stack::new();
    run2(instructions, constants, variables, functions, &mut stack);
    println!("stack max: {}", stack.max);
}

struct Stack {
    stack: Vec<i64>,
    max: usize,
}

impl Stack {
    fn new() -> Stack {
        return Stack {
            stack: Vec::with_capacity(20),
            max: 0,
        };
    }

    fn push(&mut self, value: i64) {
        self.stack.push(value);
        self.max = std::cmp::max(self.max, self.stack.len());
    }

    fn pop(&mut self) -> i64 {
        let value = self.stack.pop().unwrap();
        return value;
    }
}

fn run2(
    instructions: &[u8],
    constants: &[Constant],
    mut variables: Vec<i64>,
    functions: &[FunctionDefinition],
    mut stack: &mut Stack,
) {
    let mut ip = 0;
    while ip < instructions.len() {
        let instruction = instructions[ip];
        ip += 1;
        match instruction {
            OP_POP => {
                stack.pop();
            }
            OP_JUMP_IF_FALSE => {
                let b1 = instructions[ip] as usize;
                let b2 = instructions[ip + 1] as usize;
                ip += 2;
                let jump_amount = b1 * 256 + b2;
                let value = stack.pop();
                if value == 0 {
                    ip += jump_amount
                }
            }
            OP_JUMP_FORWARD => {
                let b1 = instructions[ip] as usize;
                let b2 = instructions[ip + 1] as usize;
                ip += 2;
                let jump_amount = b1 * 256 + b2;
                ip += jump_amount
            }
            OP_BINARY => {
                let binop = instructions[ip];
                ip += 1;
                let right = stack.pop();
                let left = stack.pop();

                match binop {
                    OP_BINARY_PLUS => stack.push(left + right),
                    OP_BINARY_SUBTRACT => stack.push(left - right),
                    OP_BINARY_MULTIPLY => stack.push(left * right),
                    OP_BINARY_DIVIDE => stack.push(left / right),
                    OP_BINARY_GREATER_THAN => stack.push(if left > right { 1 } else { 0 }),
                    OP_BINARY_LESS_THAN => stack.push(if left < right { 1 } else { 0 }),
                    _ => panic!("unknown binary operation {binop} at index {}", ip - 1),
                }
            }
            OP_INLINE_NUMBER => {
                let value = instructions[ip] as i64;
                stack.push(value);
                ip += 1;
            }
            OP_LOAD_CONSTANT => {
                let id = instructions[ip] as usize;
                let constant = &constants[id];
                match constant {
                    Constant::Number(number) => {
                        stack.push(*number);
                    }
                    Constant::String(_) => {
                        todo!("cannot handle strings yet");
                    }
                }
                ip += 1;
            }
            OP_READ_VARIABLE => {
                let id = instructions[ip] as usize;
                let value = variables[id];
                stack.push(value);
                ip += 1;
            }
            OP_SET_VARIABLE => {
                let id = instructions[ip] as usize;
                let value = stack.pop();
                variables[id] = value;
                ip += 1;
            }
            OP_PRINTLN => {
                let arg_count = instructions[ip] as usize;
                let mut values: Vec<i64> = vec![0; arg_count];
                for i in 0..arg_count {
                    values[i] = stack.pop();
                }
                values.reverse(); // TODO: reverse during iteration
                println!("println({:?})", values);
                stack.push(0); // TODO: the println return value is technically Go's "nil" (wich Toi doesn't support)
                ip += 1;
            }
            OP_CALL_FUNCTION => {
                let function_name_id = instructions[ip] as usize;
                ip += 1;
                let constant = &constants[function_name_id];
                let function_name = if let Constant::String(str) = constant {
                    str
                } else {
                    panic!("epxected string");
                };
                let function = functions.iter().find(|f| f.name == *function_name).unwrap();
                let mut function_variables: Vec<i64> = vec![0; function.variables.len()];
                for i in (0..function.parameters.len()).rev() {
                    function_variables[i as usize] = stack.pop();
                }

                run2(
                    &function.instructions,
                    constants,
                    function_variables,
                    functions,
                    &mut stack,
                );
            }
            _ => panic!("unknown instruction {instruction} at index {}", ip - 1),
        }
    }

    // TODO: should check has_out_var
    if !variables.is_empty() {
        let out_value = *variables.last().unwrap();
        stack.push(out_value);
    } else {
        stack.push(0);
    }
}

const OP_POP: u8 = 0;
const OP_BINARY: u8 = 1;
// TODO const OP_NOT: u8 = 2;
const OP_JUMP_IF_FALSE: u8 = 3;
const OP_JUMP_FORWARD: u8 = 4;
// TODO const OP_JUMP_BACK: u8 = 5;
const OP_INLINE_NUMBER: u8 = 6;
const OP_LOAD_CONSTANT: u8 = 7;
const OP_READ_VARIABLE: u8 = 8;
const OP_SET_VARIABLE: u8 = 9;
// TODO const OP_CALL_BUILTIN: u8 = 10;
const OP_CALL_FUNCTION: u8 = 11;
const OP_PRINTLN: u8 = 12;
// TODO const OP_DUPLICATE: u8 = 13;
// TODO const OP_INVALID: u8 = 14;

// OP_BINARY arguments
const OP_BINARY_PLUS: u8 = 0;
const OP_BINARY_SUBTRACT: u8 = 1;
const OP_BINARY_MULTIPLY: u8 = 2;
const OP_BINARY_DIVIDE: u8 = 3;
// TODO const OP_BINARY_REMAINDER: u8 = 4;
// TODO const OP_BINARY_EQUAL: u8 = 5;
const OP_BINARY_GREATER_THAN: u8 = 6;
const OP_BINARY_LESS_THAN: u8 = 7;
// TODO const OP_BINARY_CONCAT: u8 = 8;
